### PR TITLE
Add template linting via djlint

### PR DESCRIPTION
**Pull request by silverwind**
So I found this [linter](https://github.com/Riverside-Healthcare/djlint) which features a mode for go templates, so I gave it a try and it did find a number of valid issue, like unbalanced tags etc. It also has a number of bugs, I had to disable/workaround many issues.

Given that this linter is written in python, this does add a dependency on `python` >= 3.8 and `poetry` to the development environment to be able to run this linter locally.

- `e.g.` prefixes on placeholders are removed because the linter had a false-positive on `placeholder="e.g. cn=Search"` for the `attr=value` syntax and it's not ideal anyways to write `e.g.` into a placeholder because a placeholder is meant to hold a sample value.
- In `templates/repo/settings/options.tmpl` I simplified the logic to not conditionally create opening tags without closing tags because this stuff confuses the linter (and possibly the reader as well).

### CODE REVIEW

File: .github/workflows/files-changed.yml [hunk 1]
```diff
@@ -51,3 +55,6 @@ jobs:
 
             actions:
               - ".github/workflows/*"
+
+            templates:
+              - "**/*.tmpl"
```
Comment (delvh) [inline .github/workflows/files-changed.yml]:
```suggestion
              - "templates/**.tmpl"
```

**Comment by jolheiser** [comment]
It's nice that it found some issues to remedy, but I am very hesitant to add _another_ language ecosystem just for this.

**Comment by silverwind** [comment]
> It's nice that it found some issues to remedy, but I am very hesitant to add _another_ language ecosystem just for this.

I guess most people won't run this locally anyways, but I thought I at least enable them to do so. They don't have to install python to get gitea dev setup running.

**Comment by jolheiser** [comment]
> Erm… Have you seen how many PRs we already needed because there was an error in the templates that no one noticed?

I don't disagree that this is helpful.

> The good thing is also that Python is already installed on every machine, so you don't even need to install a new language on your system.

I'm not sure how true this is. It may be mostly true on many Linux systems, but I'm not sure about others. Furthermore, even if Python is installed, this requires installing Poetry as well, but...

> While I do agree that it would be better in a language we already support, I'll take what I can get.

I guess since it's not a requirement for dev I'm not exactly blocking, but it does feel a bit strange.

**Comment by silverwind** [comment]
> this requires installing Poetry as well

I chose Poetry because I think it's the most sane python dependency manager currently. `pipx install djlint` or `pip install djlint` will also work most likely but that leaves the problem that the dependency is unversioned so may yield differrent results depending on the installed version, so it's not something I really consider.

In short: The python dependency management ecosystems sucks, poetry is not ideal, but better then the other solutions.

**Comment by silverwind** [comment]
I just saw they support a [experimental npm installation](https://github.com/Riverside-Healthcare/djLint#-how), so we might actually be able to skip the poetry dependency (it does require python and pip still). I will try that tomorrow.

**Review by wxiaoguang — APPROVED** [review]
I think a simple "pip install" could be better if it works. Otherwise overall LGTM

File: .github/workflows/pull-compliance.yml [hunk 2]
```diff
@@ -26,6 +26,19 @@ jobs:
...
+  lint-templates:
+    if: needs.files-changed.outputs.templates == 'true'
+    needs: files-changed
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
+        with:
+          python-version: "3.11"
+      - run: pipx install poetry
```
Comment (lunny) [inline .github/workflows/pull-compliance.yml]: Since it's a one time installation, I couldn't find the difference from pip and pipx here.

Comment (silverwind) [reply .github/workflows/pull-compliance.yml]: The difference with pipx is that each applications gets it's own virtualenv, so applications can not influence each other via their dependencies. But it likely does not matter on the actions run.

I will try the npm method next, which might make this and poetry obsolete.

**Comment by silverwind** [comment]
> I think a simple "pip install" could be better if it works. Otherwise overall LGTM

It would leave the dependency unversioned, which in the long run will bring issues of inconsistent behaviour between versions, it's better to version-lock all dependencies. I will try the npm method.

**Comment by wxiaoguang** [comment]
> > I think a simple "pip install" could be better if it works. Otherwise overall LGTM
> 
> It would leave the dependency unversioned, which in the long run will bring issues of inconsistent behaviour between versions, it's better to version-lock all dependencies. I will try the npm method.

I do not think so. `pip install` can set the version. https://www.google.com/search?client=firefox-b-d&q=pip+install+with+version

**Comment by silverwind** [comment]
pip will pollute the global site-packages thought, I never use it because of that.

**Comment by wxiaoguang** [comment]
> pip will pollute the global site-packages thought, I never use it because of that.

Hmm yes, that's a problem and that's why pip is always used with virtualenv.

Anyway, I think linting the templates is good and have voted my approval 😄

**Comment by silverwind** [comment]
The npm module they publish is just a [dirty pip wrapper](https://github.com/Riverside-Healthcare/djLint/blob/485738dee4f534819d90dd3d3e34388c02b8f1c6/package.json#L34) that will pollute global python packages, so I won't consider it and think the current solution is fine as-is.

**Comment by silverwind** [comment]
https://github.com/go-gitea/gitea/pull/25212/commits/5b6a4dc7aa316b74f5dce12ee5167c8fb3b7161b changes all `form`'s method to lowercase, which is actually required as per HTML5 spec, but browsers did handle it case-insensitively. As far as I can tell, there is currently no JS that checks for these values, so it's hopefully fine. https://github.com/go-gitea/gitea/pull/25219 will check, but I'll ensure that it does so case-insensitively.

File: .github/workflows/pull-compliance.yml (hunk 2 above)
Comment (silverwind) [reply .github/workflows/pull-compliance.yml]: npm method just wraps `pip` so it not suitable. I think pipx is fine, but if you really like, I can change this to `pip` for the actions run.

**Comment by silverwind** [comment]
> I do not think so. `pip install` can set the version. https://www.google.com/search?q=pip+install+with+version
> 
> Or, use `requirements.txt` for pip (https://www.google.com/search?q=pip+requirements.txt+version) , it works like package.json or go.mod

Imho, poetry is clearly superior to `pip` in all regards. It has proper lockfiles, allows execution without "entering" the virtualenv. So I'd like to keep it. Another alternative is [pdm](https://github.com/pdm-project/pdm), but so far I don't really see much if any benefit from using it over poetry.

**Comment by delvh** [comment]
It might be the case that it is superior.
However, python is not a language that is part of the project.
I think it is a bit ridiculous to require an extra package manager for a project that doesn't use the language when there is already a version that many people have.

**Comment by silverwind** [comment]
It's only required if people want to lint templates locally. We could go the route of the global pip installation, but I forsee destroying people's other dev environments if we start polluting global site-packages.

**Comment by wxiaoguang** [comment]
Language is never a real problem IMO. And there is no hard dependency with the 3rdparty programs, no lock-in.

And this PR's approach only runs in CI environment, it won't harm anyone, so I think no need to worry about it.

File: docs/content/doc/development/hacking-on-gitea.en-us.md [hunk 3]
```diff
@@ -48,6 +48,10 @@ Gitea uses `gofmt` to format source code. However, the results of
 recommended to install the version of Go that our continuous integration is
 running. As of last update, the Go version should be {{< go-version >}}.
 
+To lint the template files, ensure [Python](https://www.python.org/) and
+[Poetry](https://python-poetry.org/) are installed. The recommended way to
+install poetry is via `pipx install poetry`.
```
Comment (silverwind) [inline docs/content/doc/development/hacking-on-gitea.en-us.md]:
```suggestion
[Poetry](https://python-poetry.org/) are installed.
```

File: .github/workflows/pull-compliance.yml (hunk 2 above)
Comment (silverwind) [inline .github/workflows/pull-compliance.yml]:
```suggestion
      - run: pip install -U poetry
```

**Comment by silverwind** [comment]
Decided to go with `pip` to install `poetry` on the action because apparently `pipx` is not provided by the `setup-python` action, but by the GitHub default environment, so the proper solution would have been to install pip to install pipx, pipx to install poetry and finally poetry to install the dependencies, but that chain was getting to insane for my tastes and Actions is mostly an isolated environment anyways where we don't need to worry so much about pip package conflicts.

File: .github/workflows/pull-compliance.yml (hunk 2 above)
Comment (silverwind) [reply .github/workflows/pull-compliance.yml]: Now using `pip`

File: .github/workflows/pull-compliance.yml [hunk 4]
```diff
@@ -26,6 +26,19 @@ jobs:
...
+  lint-templates:
+    if: needs.files-changed.outputs.templates == 'true'
+    needs: files-changed
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
+        with:
+          python-version: "3.11"
+      - run: pip install -U poetry
```
Comment (silverwind) [inline .github/workflows/pull-compliance.yml]:
```suggestion
      - run: pip install poetry
```

**Review by techknowlogick — APPROVED** [review]

File: templates/repo/issue/choose.tmpl [hunk 5]
```diff
@@ -50,7 +50,7 @@
 			<div class="ui warning message">
 				<div class="text left">
 					<div>{{.locale.Tr "repo.issues.choose.invalid_config"}}</div>
-					<diy>{{.IssueConfigError}}</div>
```
Comment (delvh) [inline templates/repo/issue/choose.tmpl:53]: :eyes:

File: templates/repo/settings/options.tmpl [hunk 6]
```diff
@@ -300,32 +296,20 @@
 				{{$isWikiEnabled := or (.Repository.UnitEnabled $.Context $.UnitTypeWiki) (.Repository.UnitEnabled $.Context $.UnitTypeExternalWiki)}}
 				<div class="inline field">
 					<label>{{.locale.Tr "repo.wiki"}}</label>
-					{{if and (.UnitTypeWiki.UnitGlobalDisabled) (.UnitTypeExternalWiki.UnitGlobalDisabled)}}
-					<div class="ui checkbox disabled" data-tooltip-content="{{.locale.Tr "repo.unit_disabled"}}">
-					{{else}}
-					<div class="ui checkbox">
-					{{end}}
+					<div class="ui checkbox{{if and (.UnitTypeWiki.UnitGlobalDisabled) (.UnitTypeExternalWiki.UnitGlobalDisabled)}} disabled{{end}}"{{if and (.UnitTypeWiki.UnitGlobalDisabled) (.UnitTypeExternalWiki.UnitGlobalDisabled)}} data-tooltip-content="{{.locale.Tr "repo.unit_disabled"}}"{{end}}>
```
Comment (delvh) [inline templates/repo/settings/options.tmpl]: Sounds like we should extract the check into a variable?

File: templates/repo/settings/options.tmpl [hunk 7]
```diff
@@ -342,22 +326,14 @@
 				{{$isIssuesEnabled := or (.Repository.UnitEnabled $.Context $.UnitTypeIssues) (.Repository.UnitEnabled $.Context $.UnitTypeExternalTracker)}}
 				<div class="inline field">
 					<label>{{.locale.Tr "repo.issues"}}</label>
-					{{if and (.UnitTypeIssues.UnitGlobalDisabled) (.UnitTypeExternalTracker.UnitGlobalDisabled)}}
-					<div class="ui checkbox disabled" data-tooltip-content="{{.locale.Tr "repo.unit_disabled"}}">
-					{{else}}
-					<div class="ui checkbox">
-					{{end}}
+					<div class="ui checkbox{{if and (.UnitTypeIssues.UnitGlobalDisabled) (.UnitTypeExternalTracker.UnitGlobalDisabled)}} disabled{{end}}"{{if and (.UnitTypeIssues.UnitGlobalDisabled) (.UnitTypeExternalTracker.UnitGlobalDisabled)}} data-tooltip-content="{{.locale.Tr "repo.unit_disabled"}}"{{end}}>
```
Comment (delvh) [inline templates/repo/settings/options.tmpl]: See above

File: templates/repo/issue/choose.tmpl (hunk 5 above)
Comment (silverwind) [reply templates/repo/issue/choose.tmpl:53]: Yeah, a quite glaring error in hindsight :D

File: templates/repo/settings/options.tmpl (hunk 6 above)
Comment (silverwind) [reply templates/repo/settings/options.tmpl]: Yeah, I will refactor these, I think there's like 10 cases in total.

File: .github/workflows/files-changed.yml (hunk 1 above)
Comment (silverwind) [reply .github/workflows/files-changed.yml]: Fixed in https://github.com/go-gitea/gitea/pull/25212/commits/4f3a54815500bf8a42000480b803671ae2c6c33c.

File: templates/repo/settings/options.tmpl (hunk 6 above)
Comment (silverwind) [reply templates/repo/settings/options.tmpl]: Refactor to vars done in https://github.com/go-gitea/gitea/pull/25212/commits/c523ca33115553aae943fabf5b67bbb0fa560d2e.

File: templates/repo/settings/options.tmpl [hunk 8]
```diff
@@ -285,47 +285,35 @@
...
+					<div class="ui checkbox{{if $isCodeGlobalDisabled}} disabled{{end}}"{{if $isCodeGlobalDisabled}} data-tooltip-content="{{.locale.Tr "repo.unit_disabled"}}"{{end}}>
 						<input class="enable-system" name="enable_code" type="checkbox"{{if $isCodeEnabled}} checked{{end}}>
 						<label>{{.locale.Tr "repo.code.desc"}}</label>
 					</div>
 				</div>
 
 				{{$isWikiEnabled := or (.Repository.UnitEnabled $.Context $.UnitTypeWiki) (.Repository.UnitEnabled $.Context $.UnitTypeExternalWiki)}}
+				{{$isWikiGlobalDisabled := .UnitTypeWiki.UnitGlobalDisabled}}
+				{{$isExternalWikiGlobalDisabled := .UnitTypeExternalWiki.UnitGlobalDisabled}}
+				{{$isBothWikiGlobalDisabled := and $isWikiGlobalDisabled $isBothWikiGlobalDisabled}}
```
Comment (delvh) [inline templates/repo/settings/options.tmpl:300]: Regression. You meant `isExternalWikiGlobalDisabled`

Comment (silverwind) [reply templates/repo/settings/options.tmpl:300]: Ah, indeed. Fix coming.

Comment (silverwind) [reply templates/repo/settings/options.tmpl:300]: BTW I really wonder how go does not complain about `var := var` here. It definitely should.

Comment (silverwind) [reply templates/repo/settings/options.tmpl:300]: https://github.com/go-gitea/gitea/pull/25267

**Pull request merged**